eos-boot-upgrade

```diff
@@ -0,0 +1,278 @@
+    options = {}
+    with open('/proc/cmdline', 'r') as f:
+        for option in f.readline().strip().split():
+            fields = option.split("=", 1)
+            if len(fields) == 1:
+                options[fields[0]] = True
+            else:
+                options[fields[0]] = fields[1]
+    return options
+
+
+def get_disk():
+    '''Get the booting disk'''
+    disk = Disk.get_booting_device()
+    print('[*] Booting disk: ' + disk.device)
+
+    if disk.label != 'gpt':
+        raise BootUpgradeException('[!] Partition table is not GPT')
+
+    root_partition = disk.get_partition_by_guid('4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709')
+    if not root_partition:
+        raise BootUpgradeException('[!] root partition not found (or wrong GUID)')
+
+    print('[*] Root partition: ' + root_partition.device)
+    return disk
+
+
+def get_esp(disk):
+    '''Get the ESP partition data'''
+    esp_partition = disk.get_partition_by_guid('C12A7328-F81F-11D2-BA4B-00A0C93EC93B')
+    if not esp_partition:
+        raise BootUpgradeException('[!] ESP partition not found (or wrong GUID)')
+    if not esp_partition.device.endswith('1'):
+        raise BootUpgradeException('[!] ESP is not the first partition')
+
+    print('[*] ESP partition: ' + esp_partition.device)
+    return esp_partition
+
+
+def get_dir_size(dir_path):
+    '''Get the directory size'''
+    total_size = 0
+    for (dirpath, _, filenames) in os.walk(dir_path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += os.path.getsize(fp)
+    return total_size
+
+
+def check_esp_space(esp_mount, ostree_boot_efi_path):
+    '''Check that we have enough space in the ESP to copy the new data'''
+    ostree_boot_efi_size = get_dir_size(ostree_boot_efi_path)
+    esp_free_space = shutil.disk_usage(esp_mount).free
+    return esp_free_space > ostree_boot_efi_size
+
+
+def dir_cmp(dcmp):
+    '''Helper for directories comparison'''
+    if dcmp.diff_files or dcmp.right_only:
+        return True
+
+    for sub_dcmp in dcmp.subdirs.values():
+        if dir_cmp(sub_dcmp):
+            return True
+    return False
+
+
+def check_for_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Recursively compare the files to upgrade'''
+    dcmp = filecmp.dircmp(esp_mount, ostree_boot_efi_path)
+    return dir_cmp(dcmp)
+
+
+def do_uefi_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Actually start the UEFI upgrade copying the files'''
+    esp_mount_tmp = os.path.join(esp_mount, 'EFI.new')
+
+    print('[*] Updated:')
+    try:
+        file_list = distutils.dir_util.copy_tree(ostree_boot_efi_path,
+                                                 esp_mount_tmp)
+        os.sync()
+
+        for f_source in file_list:
+            f_rp = os.path.relpath(f_source, esp_mount_tmp)
+            f_dest = os.path.join(esp_mount, f_rp)
+
+            if not os.path.isfile(f_dest) or not filecmp.cmp(f_source, f_dest):
+                shutil.move(f_source, f_dest)
```

Comment:
`renameat2(..., RENAME_EXCHANGE)` doesn't work on FAT, so we can't use that to swap the old/new `EFI\Endless` directories atomically. Without some mechanism to juggle a pair of BootOrder entries we can't entirely avoid this window between updating the files. So we need to:

1. make the window as narrow as possible
2. add/replace files in a carefully-determined order to reduce the number of ways that a poorly-timed loss of power could mess us up

Here are two examples of how the order can affect whether the system boots after a loss of power.

Suppose we have to rotate the key our GRUB is signed with. Then suppose while running this script the computer loses power in between overwriting Shim with a new version that doesn't trust the old key, and then overwriting GRUB with a new version signed with a new key. Then the system won't boot in secure boot mode. If we made sure any new GRUB binaries were still signed with the old key as well as the new one, then if we always copied the new GRUB executable before copying the new Shim, the old Shim would still trust the new GRUB.

Another example: before eos3.1, the ESP contained:

* \EFI\BOOT\
  * bootx64.efi (shim)
  * grubx64.efi

From eos3.1:

* \EFI\BOOT\
  * bootx64.efi (shim)
  * fallback.efi
* \EFI\Endless\
  * shim.efi
  * grubx64.efi
  * BOOT.CSV

When upgrading from the former to the latter we need to take care to update the full contents of \EFI\Endless *before* we add fallback.efi, because the logic in Shim is: "if (run as \EFI\BOOT\BOOT*.EFI && \EFI\BOOT\fallback.efi exists) { run fallback } else { run \EFI\BOOT\grubx64.efi }", so if we add fallback.efi before \EFI\Endless is fully-populated, and then lose power, the system will no longer boot.